Add support for amp-web-push

jasonpang opened the pull request:
This PR implements the first version of `amp-web-push`.

We'll be adding more tests and type information comments, but we'd like to begin the code review process.

Design Review: [Document](https://docs.google.com/document/d/15VBR9Ht373TW0KwYD5RUGfafX4nELbP-6uWj7ngIRWU/)
Intent to Implement: #8791

jasonpang commented on examples/amp-web-push.amp.html:
```diff
@@ -0,0 +1,71 @@
+<!doctype html>
```
For this commit, I changed the committer email to match that listed on the Google CLA.

jasonpang commented:
@googlebot I signed it!

cramforce commented on examples/amp-web-push.service-worker.js:
```diff
@@ -0,0 +1,81 @@
+import { WorkerMessenger } from '../extensions/amp-web-push/0.1/worker-messenger';
```
Copyright note, please

jasonpang replied:
Added.

cramforce commented on extensions/amp-web-push/0.1/amp-web-push-widget.js:
```diff
@@ -0,0 +1,54 @@
+ * the License.
+ */
+
+import { TAG } from './vars';
```
nit: no whitespace around `TAG`

jasonpang replied:
Removed all whitespace around `import` calls.

cramforce commented on extensions/amp-web-push/0.1/amp-web-push-widget.js:
```diff
@@ -0,0 +1,54 @@
+
+
+export const WebPushWidgetVisibilities = {
+  /**
```
nit: only `**` for jsdoc

jasonpang replied:
I changed `/**` to `/*` wherever there wasn't "real" JSDoc (no `@something). I plan to go back and add back JSDoc for public methods (methods that aren't underscored `_`). For now, the methods just have regular multi-line comments.

cramforce commented on extensions/amp-web-push/0.1/amp-web-push-widget.js:
```diff
@@ -0,0 +1,54 @@
+import { TAG } from './vars';
+import { Layout } from '../../../src/layout';
+
+
```
`/** @enum {string} */`

jasonpang replied:
Added the `@enum` type comments to all the enums.

cramforce commented on extensions/amp-web-push/0.1/iframe.js:
```diff
@@ -0,0 +1,52 @@
+import { setStyle } from '../../../src/style';
+
+export default class IFrame {
+  constructor(document, url) {
```
Please add JSDoc. Let me know if you need references.

jasonpang replied:
Added comments to the IFrame class.

cramforce commented on extensions/amp-web-push/0.1/vars.js:
```diff
@@ -0,0 +1,33 @@
+ * the License.
+ */
+
+ /** @const */
```
Nit: redundant

jasonpang replied:
Removed the redundant `@const` markers in this file.

cramforce commented on extensions/amp-web-push/0.1/iframe.js:
```diff
@@ -0,0 +1,52 @@
+
+    this.document.body.appendChild(this.domElement);
+
+    this.loadPromise = new Promise((resolve, reject) => {
```
Can you use `event-helper.js` `loadPromise`?

jasonpang replied:
Replaced my `loadPromise` implementation with `event-helper.js`'s `loadPromise`. I had to [whitelist](https://github.com/ampproject/amphtml/pull/10468/commits/63f8b5e096e688b4a7aab2c134c00b26977d2a4b) `iframe.js` though.

cramforce commented on extensions/amp-web-push/0.1/iframe.js:
```diff
@@ -0,0 +1,52 @@
+
+import { setStyle } from '../../../src/style';
+
+export default class IFrame {
```
I didn't find where this was used.

jasonpang replied:
This is used as a helper abstraction when [creating the IFrame](https://github.com/OneSignal/amp-web-push/blob/63f8b5e096e688b4a7aab2c134c00b26977d2a4b/extensions/amp-web-push/0.1/web-push-service.js#L111) and when [loading the IFrame](https://github.com/OneSignal/amp-web-push/blob/63f8b5e096e688b4a7aab2c134c00b26977d2a4b/extensions/amp-web-push/0.1/web-push-service.js#L119).

cramforce commented on extensions/amp-web-push/0.1/worker-messenger.js:
```diff
@@ -0,0 +1,173 @@
+  }
+}
+
+class WorkerMessenger {
```
Please add some comments. Need to understand why such a complicated thing is needed.

jasonpang replied:
I decided to [remove `worker-messenger.js`](https://github.com/ampproject/amphtml/pull/10468/commits/47a45270558ac18d7d2c53ae5551209eb92be2a6) in the end.

The bulk of the code was around the abstraction of providing a `once()` listener, so that you could do:

`workerMessenger.once('some-topic', myCallback())`

I was able to find a reasonable workaround for it. I took out the worker messenger from the extension + the sample file that used to use it.

honeybadgerdontcare commented:
I noticed that the extension is called `amp-web-push` but the tag is `amp-web-push-widget`. Is there a reason for that?

jasonpang commented:
@honeybadgerdontcare This is now changed so that `<amp-web-push>` is the only custom element name. `<amp-web-push-widget>` no longer exists.

In addition to the original `<amp-web-push-widget>` custom element, I also had a service registered as `'amp-web-push'`. The service used to expose custom actions `'subscribe'` and `'unsubscribe'` (used as: `tap:amp-web-push.subscribe`). I had initially kept the names separate, but looking back, I could have also registered the custom element with the same name `'amp-web-push'` as the service.

honeybadgerdontcare commented on extensions/amp-web-push/0.1/validator-amp-web-push.protoascii:
```diff
@@ -0,0 +1,42 @@
+}
+tags: {  # <amp-web-push>
+  html_format: AMP
+  tag_name: "amp-web-push"
```
This needs to be capitalized: `tag_name: "AMP-WEB-PUSH"`

honeybadgerdontcare commented on extensions/amp-web-push/0.1/validator-amp-web-push.protoascii:
```diff
@@ -0,0 +1,42 @@
+tags: {  # amp-web-push
+  html_format: AMP
+  tag_name: "SCRIPT"
+  satisfies: "amp-web-push extension .js script"
```
`satisfies` is not needed any more

honeybadgerdontcare commented on extensions/amp-web-push/0.1/validator-amp-web-push.protoascii:
```diff
@@ -0,0 +1,42 @@
+tags: {  # <amp-web-push>
+  html_format: AMP
+  tag_name: "amp-web-push"
+  requires: "amp-web-push extension .js script"
```
`requires` is not needed any more

honeybadgerdontcare replied:
apologies, `requires` has been replaced by `requires_extension` which in this case would be `requires_extension: "amp-web-push"`.

honeybadgerdontcare commented on extensions/amp-web-push/0.1/validator-amp-web-push.protoascii:
```diff
@@ -0,0 +1,42 @@
+#
```
We've moved validator protoasciis up one level, outside of the version directory. E.g. `extensions/amp-web-push/validator-amp-web-push.protoascii`.

dvoytenko commented on extensions/amp-web-push/0.1/web-push-service.js:
```diff
@@ -0,0 +1,697 @@
+    this.popupMessenger_ = null;
+  }
+
+  /*
```
`/**`

jasonpang replied:
Fixed JSDoc /** asterisk count.

dvoytenko commented on extensions/amp-web-push/0.1/web-push-service.js:
```diff
@@ -0,0 +1,697 @@
+
+  /*
+  * Occurs when the DOM is ready to be parsed.
+  */
```
`@private`

jasonpang replied:
Changed private methods to @private and underscored.

dvoytenko commented on extensions/amp-web-push/0.1/web-push-service.js:
```diff
@@ -0,0 +1,697 @@
+      // An error will already be thrown from the config parsing function
+      return;
+    }
+    this.debug_ = this.config_.debug;
```
We should just enable debug via `getMode().development`. All similar components do that.

jasonpang replied:
Set `debug: getMode().development` for the main `web-push-service.js` file. 

Did not change `helper-frame.js` and `permission-dialog.js` since these files are outside the AMP framework and live on the canonical origin. `window-messenger.js` inherits the debug setting passed to it.

dvoytenko commented on extensions/amp-web-push/0.1/web-push-service.js:
```diff
@@ -0,0 +1,697 @@
+      `parentOrigin=${window.location.origin }`;
+
+    // Create a hidden iFrame to check subscription state
+    this.iframe_ = new IFrame(this.ampdoc.win.document, finalIframeUrl);
```
It's best to pass `this.ampdoc` and then use `this.ampdoc.getBody().appendChild()` and such.

jasonpang replied:
Passed `ampdoc` to IFrame.

dvoytenko commented on extensions/amp-web-push/0.1/web-push-service.js:
```diff
@@ -0,0 +1,697 @@
+
+  log_(text) {
+    // Only prints if the user turned on debug/verbose mode
+    if (this.debug_) {
```
This should just be `user().fine()` or similar. `console.log` is not allowed.

jasonpang replied:
Replaced all `console.log()` statements with `dev().fine()`.

dvoytenko commented on extensions/amp-web-push/0.1/web-push-service.js:
```diff
@@ -0,0 +1,697 @@
+   */
+  ensureAmpExperimentEnabled_() {
+    // Allow integration tests to access this class in testing mode.
+    /** @const @private {boolean} */
```
Redundant.

jasonpang replied:
Removed JSDoc reference for `isExperimentEnabled`.

dvoytenko commented on extensions/amp-web-push/0.1/web-push-service.js:
```diff
@@ -0,0 +1,697 @@
+    return configJsonNode.textContent;
+  }
+
+  /*
```
`/**` + `@param`

jasonpang replied:
Added /** JSDoc.

dvoytenko commented on extensions/amp-web-push/0.1/web-push-service.js:
```diff
@@ -0,0 +1,697 @@
+    }
+
+    if (!config['serviceWorkerUrl'] ||
+      new URL(config['serviceWorkerUrl']).protocol !== 'https:') {
```
Please use `url.js` `parseUrl`. Still enough user agents not supporting URL API.

jasonpang replied:
Replaced all `new URL()` with `parseUrl()`.

dvoytenko commented on extensions/amp-web-push/0.1/web-push-service.js:
```diff
@@ -0,0 +1,697 @@
+        'to be installed.');
+    }
+
+    if (new URL(config['serviceWorkerUrl']).origin !==
```
ditto

jasonpang replied:
Replaced.

dvoytenko commented on extensions/amp-web-push/0.1/web-push-service.js:
```diff
@@ -0,0 +1,697 @@
+
+  isValidHelperOrPermissionDialogUrl_(url) {
+    try {
+      const parsedUrl = new URL(url);
```
ditto

dvoytenko commented on extensions/amp-web-push/0.1/web-push-service.js:
```diff
@@ -0,0 +1,697 @@
+  isServiceWorkerActivated_() {
+    const self = this;
+    return this.queryServiceWorkerState_().then(
+      /** @param {{isControllingFrame:boolean, state:string, url:string}}
```
Ideally this would be defined in the `queryServiceWorkerState_`'s jsdoc instead of here.

jasonpang replied:
Moved JSDoc.

dvoytenko commented on extensions/amp-web-push/0.1/web-push-service.js:
```diff
@@ -0,0 +1,697 @@
+          serviceWorkerState.state === 'activated';
+
+          return isControllingFrame &&
+          serviceWorkerHasCorrectUrl &&
```
nit: indent

jasonpang replied:
Fixed indent.

dvoytenko commented on extensions/amp-web-push/0.1/web-push-service.js:
```diff
@@ -0,0 +1,697 @@
+  updateWidgetVisibilities() {
+    return this.queryNotificationPermission_().then(notificationPermission => {
+      if (notificationPermission === NotificationPermission.DENIED) {
+        this.updateWidgetVisibilitiesNotificationPermissionsBlocked();
```
Would it be ok if all of these methods were `@private`?

jasonpang replied:
May need to make these methods public for testing, so leaving for now. But will make `@private` if they aren't used for testing.

dvoytenko replied:
SG. Generally, though, I'm very much ok with using private methods for testing, especially if they are documented as such.

dvoytenko commented on extensions/amp-web-push/0.1/web-push-service.js:
```diff
@@ -0,0 +1,697 @@
+  }
+
+  updateWidgetVisibilitiesNotificationPermissionsBlocked() {
+    this.setWidgetVisibilities(WebPushWidgetVisibilities.UNSUBSCRIBED, false);
```
Is it always one-of? If so, would be probably be better to do

```
set?State(flag) {
  for (const k in enum) {
    this.setWidgetVis(enum[k], enum[k] == flag);
  }
}
```

?

jasonpang replied:
I replaced a couple redundant lines with an existing function, but I wasn't able to figure out a further optimization.

dvoytenko commented on extensions/amp-web-push/0.1/web-push-service.js:
```diff
@@ -0,0 +1,697 @@
+            // User blocked
+              reply({closeFrame: true});
+              return this.updateWidgetVisibilities();
+              break;
```
No `break` after `return`

jasonpang replied:
Removed redundant `return` statements after break and throw statements.

dvoytenko commented on extensions/amp-web-push/0.1/web-push-service.js:
```diff
@@ -0,0 +1,697 @@
+      numbers below are just reasonable defaults.
+    */
+    return {
+      width: 650,
```
Probably would be good to minmax with `screen.width/height`

jasonpang replied:
Found and copied min/max code from `amp-access` opening the dialog.

dvoytenko commented on extensions/amp-web-push/0.1/web-push-service.js:
```diff
@@ -0,0 +1,697 @@
+        this.ampdoc.win,
+        openingPopupUrl,
+        '_blank',
+        'scrollbars=yes, width=' +
```
Feel free to use string interpolations.

jasonpang replied:
Found and copied some existing organized code from `amp-access`.

dvoytenko commented on extensions/amp-web-push/0.1/web-push-service.js:
```diff
@@ -0,0 +1,697 @@
+  resumeSubscribingForPushNotifications() {
+    // Remove the ?amp-web-push-subscribing=true from the URL without affecting
+    // the page contents using the History API
+    window.history.replaceState(
```
`this.ampdoc.win.history.replaceState()`

jasonpang replied:
Changed to this.ampdoc.win.

dvoytenko commented on extensions/amp-web-push/0.1/web-push-service.js:
```diff
@@ -0,0 +1,697 @@
+            case NotificationPermission.DENIED:
+            // User blocked
+              return this.updateWidgetVisibilities();
+              break;
```
ditto

jasonpang replied:
Removed redundant break;

dvoytenko commented on extensions/amp-web-push/0.1/web-push-service.js:
```diff
@@ -0,0 +1,697 @@
+    The helper iframe HTTPS is enforced when checking the configuration.
+   */
+  isAmpPageHttps() {
+    return location.protocol === 'https:';
```
`this.ampdoc.win.location...`

jasonpang replied:
Changed to this.ampdoc.win.

dvoytenko commented on extensions/amp-web-push/0.1/window-messenger.js:
```diff
@@ -0,0 +1,383 @@
+ * the License.
+ */
+
+ /*
```
`/** @fileoverview`

jasonpang replied:
Added JSDoc and @fileoverview.

dvoytenko commented on extensions/amp-web-push/0.1/window-messenger.js:
```diff
@@ -0,0 +1,383 @@
+   * Set debug to true to get console logs anytime a message is received,
+   * sent, or discarded.
+   */
+  constructor(options) {
```
ditto: jsdoc + private

jasonpang replied:
Added JSDoc. Should this constructor be private though?

dvoytenko replied:
We can't mark constructors as private in Closure, I don't believe. Although I could be mistaken.

dvoytenko commented on extensions/amp-web-push/0.1/window-messenger.js:
```diff
@@ -0,0 +1,383 @@
+        /** @type {(function (Event): (boolean|undefined)|null)} */
+        (this.onListenConnectionMessageReceivedProc));
+      if (this.debug) {
+        console/*OK*/.log('Listening for a connection message...');
```
`dev().fine()` or `user().fine()`

jasonpang replied:
Replaced all `console.log()` statements with `dev().fine()`.

jasonpang commented on extensions/amp-web-push/0.1/iframe.js:
```diff
@@ -0,0 +1,52 @@
+ * that resolves a Promise when the iFrame has finished loading.
+ */
+export class IFrame {
+  constructor(document, url) {
```
Added JSDoc on params and vars for `iframehost.js`.

jasonpang commented on extensions/amp-web-push/0.1/iframe.js:
```diff
@@ -0,0 +1,52 @@
+ * Wraps the creation of an invisible sandboxed IFrame. Exposes a load() method
+ * that resolves a Promise when the iFrame has finished loading.
+ */
+export class IFrame {
```
Changed to `IFrameHost`.

jasonpang commented on extensions/amp-web-push/0.1/amp-web-push.js:
```diff
@@ -0,0 +1,33 @@
+import {CSS} from '../../../build/amp-web-push-0.1.css';
+
+
+AMP.registerServiceForDoc(TAG, WebPushService);
```
Changed the service name to 'amp-web-push-service'. The widget is now 'amp-web-push-widget'. The config element (replacing JSON script-based config) is now 'amp-web-push-config'.

jasonpang commented on extensions/amp-web-push/0.1/amp-web-push.js:
```diff
@@ -0,0 +1,33 @@
+ * the License.
+ */
+
+/*
```
Added `@fileoverview` to this and all other files.

jasonpang commented on extensions/amp-web-push/0.1/amp-web-push.css:
```diff
@@ -0,0 +1,47 @@
+  padding: 10px 15px;
+  cursor: pointer;
+  outline: none;
+  font-size: inherit;
```
This CSS is now removed because we think it's better to provide styling as an example instead of as a default.

jasonpang commented on extensions/amp-web-push/0.1/amp-web-push-widget.js:
```diff
@@ -0,0 +1,77 @@
+        this.onUnsubscribe.bind(this));
+  }
+
+  onSubscribe() {
```
Went around and marked `_` (underscore) methods as private. Haven't gotten to all of them / some may change because some may become public again for testing.

jasonpang commented on extensions/amp-web-push/0.1/amp-web-push-widget.js:
```diff
@@ -0,0 +1,77 @@
+  /** @override */
+  buildCallback() {
+    // Hide the element
+    this.element.classList.add('invisible');
```
Changed `.invisible` to `.amp-invisible`.

jasonpang commented on examples/amp-web-push/publisher-files/permission-dialog.js:
```diff
@@ -0,0 +1,116 @@
+   * @param {string=} opt_fallback
+   * @return {string}
+   */
+  tryDecodeUriComponent(component, fallback = '') {
```
Yep! Now that we have a gulp build task, the function is imported from the AMP framework.

jasonpang commented on examples/amp-web-push/publisher-files/permission-dialog.html:
```diff
@@ -0,0 +1,64 @@
+  </script>
+</head>
+<body>
+  <p class="centered" id="subscribe-message">Click Allow to subscribe.</p>
```
We actually think it's best to ship the extension without this text at all, and instead provide users a working example of what they could add. That way, we're not encouraging any specific kind of implementation. For the sample CSS we have in amp-web-push.css, we'd like to take that out as well and provide it as an example implementation, but not ship the base plugin with this default style.

jasonpang commented on examples/amp-web-push/publisher-files/helper-frame.html:
```diff
@@ -0,0 +1,60 @@
+<head>
+  <meta charset="utf-8">
+  <!--
+    Minified script bundle of helper-frame.js, window-messenger.js, and
```
Yes, basically this file can be copy pasted as is. 

Except now its no longer in `examples`, but it's back in the `extensions/amp-web-push/publisher-files` directory (had some trouble with the `gulp` task and it was easier to put it here).

I used to minify it using UglifyJS2's online demo, but that won't scale so we now have a gulp task that runs with `gulp dist` to produce the minified HTML publisher files.

jasonpang commented on examples/amp-web-push/publisher-files/helper-frame.js:
```diff
@@ -0,0 +1,277 @@
+  }
+
+  getParentOrigin() {
+    const queryParams = this.parseQueryString(window.location.search);
```
Will look into this as I add more tests.

jasonpang commented on examples/amp-web-push/publisher-files/helper-frame.js:
```diff
@@ -0,0 +1,277 @@
+      this.allowedWorkerMessageTopics[message.topic] = resolve;
+
+      // The AMP message is forwarded to the service worker
+      return this.waitUntilWorkerControlsPage().then(() => {
```
I removed the unnecessary return, but this Promise resolves only when `resolve()` is called (which is stored to be called later). So the return statements inside this `new Promise() { .. }` don't affect the resolution of the Promise.

jasonpang commented on examples/amp-web-push/publisher-files/helper-frame.js:
```diff
@@ -0,0 +1,277 @@
+ */
+
+class AmpWebPushHelperFrame {
+  constructor(options) {
```
Added JSDocs.

These files were originally minified by using the Typescript playground site (www.typescriptlang.org/play/) and then the UglifyJS online demo. The new gulp build process removes this manual copying.

jasonpang commented on examples/amp-web-push/publisher-files/amp-web-push.service-worker.js:
```diff
@@ -0,0 +1,135 @@
+/*
+  Broadcasts a single boolean describing whether the user is subscribed.
+ */
+async function onMessageReceived_SubscriptionState(payload) {
```
I replaced the async/await with regular Promise chains.

jasonpang commented on examples/amp-web-push/publisher-files/amp-web-push.service-worker.js:
```diff
@@ -0,0 +1,135 @@
+  }
+});
+
+/*
```
Fixed the method doc comments in this file to be JSDoc comments.

jasonpang commented on examples/amp-web-push/publisher-files/amp-web-push.service-worker.js:
```diff
@@ -0,0 +1,135 @@
+    - payload: An optional JavaScript object containing extra data relevant to
+      the command.
+   */
+  const { command, payload } = event.data;
```
Fixed by `gulp lint --fix` :).

jasonpang commented on examples/amp-web-push/publisher-files/amp-web-push.service-worker.js:
```diff
@@ -0,0 +1,135 @@
+  /*
+    Used to request the current subscription state.
+   */
+  AMP_SUBSCRIPION_STATE: "amp-web-push-subscription-state",
```
Linted this file and fixed all the errors, including single quotes.

jasonpang commented on examples/amp-web-push/publisher-files/amp-web-push.service-worker.js:
```diff
@@ -0,0 +1,135 @@
+ */
+
+ /*
+  This file is an example implementation for a service worker compatible with
```
Checked and added @fileoverview to the file JSDocs I could find.

jasonpang commented on examples/amp-web-push.amp.html:
```diff
@@ -0,0 +1,71 @@
+
+  <!-- AMP Web Push Extension Configuration
+  -->
+  <script id="amp-web-push" type="application/json">
```
I replaced the `amp-access`-like JSON-based config with your suggested element-based config.

I initially went with the JSON-based config because there was an example from `amp-access`.

jasonpang commented on build-system/tasks/presubmit-checks.js:
```diff
@@ -647,19 +647,21 @@ var forbiddenTermsSrcInclusive = {
       '3p/integration.js',
       'dist.3p/current/integration.js',
       'examples/pwa/pwa.js',
+      'examples/amp-web-push/publisher-files/helper-frame.js',
```
`tryDecodeUriComponent()` is now imported from the AMP framework instead of duplicated.

I added a `gulp build-web-push-publisher-files` build task for generating publisher files that's run with `gulp dist`. So I'm now able to import AMP framework methods in to the two publisher files `helper-frame.js` and `permission-dialog.js` without duplicating the code.

dvoytenko replied:
/cc @erwinmombay for gulpfile changes.

Cool. But can you explain this a bit more how you see this? Do you want `publisher-files` to be in the GitHub master so that devs can just pick them up as examples? Or do you want them to be deployed somewhere on CDN and instruct devs to simply proxy them?

dvoytenko commented on extensions/amp-web-push/0.1/permission-dialog.js:
```diff
@@ -0,0 +1,121 @@
+   * @param {string} queryString
+   * @return {!Object<string>}
+   */
+  parseQueryString(queryString) {
```
From `url.js`?

jasonpang replied:
Replaced duplicated framework method with framework import.

dvoytenko commented on extensions/amp-web-push/0.1/web-push-service.js:
```diff
@@ -0,0 +1,624 @@
+      installStylesForShadowRoot(root, CSS, false, TAG);
+    }
+
+    /** @private {Object} */
```
`!Object`

jasonpang replied:
Made `this.config` non-nullable with `!` type annotation.

dvoytenko commented on extensions/amp-web-push/0.1/web-push-service.js:
```diff
@@ -0,0 +1,624 @@
+   * @param {!../../../src/service/ampdoc-impl.AmpDoc} ampdoc
+   */
+  constructor(ampdoc) {
+    this.ampdoc = ampdoc;
```
`/** @const */`

jasonpang replied:
Added `@const` to `this.ampdoc`.

dvoytenko commented on extensions/amp-web-push/0.1/web-push-service.js:
```diff
@@ -0,0 +1,624 @@
+
+  /** @private */
+  isContinuingSubscriptionFromRedirect_() {
+    return location.search.indexOf(
```
Is this `window.location`? If so, please use `ampdoc.win.location`

jasonpang replied:
Changed to use `ampdoc.win` instead of `window` / a test location when testing.

dvoytenko commented on extensions/amp-web-push/0.1/web-push-service.js:
```diff
@@ -0,0 +1,624 @@
+    );
+  }
+
+  /** @private */
```
`@return` here and below.

jasonpang replied:
Added JSDoc.

dvoytenko commented on extensions/amp-web-push/0.1/web-push-service.js:
```diff
@@ -0,0 +1,624 @@
+   * This action is exposed from this service and is called from the
+   * <amp-web-push> custom element.
+   *
+   * @public
```
Public implied.

jasonpang replied:
Removed all `@public` JSDoc annotations, since they are by default.

dvoytenko commented on extensions/amp-web-push/0.1/web-push-service.js:
```diff
@@ -0,0 +1,624 @@
+    @private
+   */
+  isAmpPageHttps_() {
+    return location.protocol === 'https:';
```
Ditto, `ampdoc.win.location`

jasonpang replied:
Replaced remaining `window.` in class with `ampdoc.win`.

dvoytenko commented on extensions/amp-web-push/0.1/window-messenger.js:
```diff
@@ -0,0 +1,397 @@
+     * describing incoming replies and outgoing sends. Just used to internally
+     * keep track of replies and sends.
+     */
+    this.messages = {};
```
All/most of these are `@private`, right? It matters, b/c all `@private` are minified.

jasonpang replied:
Fixed relevant `window-messenger.js` variables and methods to be private with underscore.

jasonpang commented on examples/amp-web-push.amp.html:
```diff
@@ -0,0 +1,69 @@
+
+  <!-- AMP Web Push Extension Configuration
+  -->
+  <amp-web-push-config
```
Thanks for the detailed explanation!

We moved `<amp-web-push-config>` from `<head>` to `<body>`.

We believe it's better to keep `<amp-web-push>` as the main widget element and `<amp-web-push-config>` as a unique configuration-only element.

We don't think the first solution is suitable because we'd like publishers to be able to add `<amp-web-push>` widgets to multiple places in their document, rather then requiring them to be nested under a single `<amp-web-push>` element.

In your second suggestion the configuration attributes' repetitions seems less DRY and more error prone for the attributes that need to be repeated. Also, using `<amp-web-push-config id="amp-web-push">` allows all tap actions to be "tap:amp-web-push.action", instead of "tap:widget-id.action". This feels more intuitive because there's only one instance of web push the user is subscribing to or unsubscribing from.

dvoytenko replied:
Ok. Cool. One more question: what do you think about calling config `<amp-web-push>` and widgets `<amp-web-push-widget>`? This would make things a bit more natural in your examples, imho.

jasonpang commented on examples/amp-web-push.amp.html:
```diff
@@ -0,0 +1,69 @@
+  <!-- AMP Web Push Extension Configuration
+  -->
+  <amp-web-push-config
+    helperIframeUrl="https://example.com/helper-iframe.html"
```
Changed the attribute names from pascalCase to be dash separated.

jasonpang commented on extensions/amp-web-push/0.1/amp-web-push-config.js:
```diff
@@ -0,0 +1,204 @@
+   * @private
+   */
+  ensureSpecificElementId_() {
+    if (this.element.getAttribute('id') !== TAG) {
```
I think this is necessary. This is supposed to check that any created <amp-web-push-config> element has only one allowed ID "amp-web-push", so that tap actions can be routed to this config element (if we are keeping the `<amp-web-push-config>` element).

jasonpang commented on extensions/amp-web-push/0.1/amp-web-push-config.js:
```diff
@@ -0,0 +1,204 @@
+   * @private
+   */
+  ensureUniqueElement_() {
+    if (this.getAmpDoc().win.document.querySelectorAll('#' + TAG).length > 1) {
```
I actually had unit tests using `getAmpDoc()` throwing error `no ampdoc, not attached`, so I ended up using `this.win.document.body`, except actually using `scopedQuerySelectorAll()`.

dvoytenko replied:
It should be called at the right time, e.g. in or after `buildCallback` and it'll work then. You have to use root node in ampdoc, b/c otherwise querying `win.document.body` will not work. E.g. in case when an ampdoc is in a shadow root, `win.document.querySelector()` won't be able to penetrate shadow root to look there. Let's find a way to get `getAmpDoc()` working here.

jasonpang commented on extensions/amp-web-push/0.1/amp-web-push-widget.js:
```diff
@@ -0,0 +1,60 @@
+
+/**
+ * @fileoverview
+ * A <web-push-widget> that shows or hides based on the page user's subscription
```
We think <amp-web-push> is the best widget name available and `amp-web-push-widget.js` is the best file name available considering potential naming conflicts.

It looks like the convention is to use `amp-web-push-impl.js`, but that name is misleading because our implementation is really split across `amp-web-push-service.js`, `helper-frame.js`, and `permission-dialog.js`.

Since our extension is called `amp-web-push`, I think the most used element, the widget itself, should be `<amp-web-push>`, with an additional one-time `<amp-web-push-config>` element.

dvoytenko replied:
Ditto here: I think calling them `amp-web-push` and `amp-web-push-widget` would clear up all these things.

jasonpang commented on extensions/amp-web-push/0.1/amp-web-push.css:
```diff
@@ -0,0 +1,19 @@
+ * the License.
+ */
+
+.amp-invisible {
```
Whoops, good catch! Made it `amp-web-push.amp-invisible`.

jasonpang commented on extensions/amp-web-push/0.1/amp-web-push.service-worker.js:
```diff
@@ -0,0 +1,155 @@
+function onMessageReceivedSubscribe() {
+  self.registration.pushManager.subscribe({
+    userVisibleOnly: true,
+    applicationServerKey: 'fake-demo-key',
```
Added a detailed comment recommending using existing subscription code for users that have it, and otherwise linking to guides on how to obtain an `applicationServerKey`. Ditto for the `unsubscribe()` method later down.

jasonpang commented on extensions/amp-web-push/0.1/amp-web-push.service-worker.js:
```diff
@@ -0,0 +1,155 @@
+  self.registration.pushManager.getSubscription()
+      .then(subscription => subscription.unsubscribe())
+      .then(() => {
+    // OPTIONALLY IMPLEMENT: Forward the unsubscription to your server here
```
This is a marker for the reader reviewing a service worker implementation example. Users using push vendors shouldn't have to deal with this, as the vendor should be implementing this. For users with their own service worker, they'll most likely already have code subscribing and unsubscribing users, and they'll just take away the parts relevant to AMP (the messages accepted by the service worker and the kinds of replies to broadcast). The code for subscribing and unsubscribing users is more to show how and where the pieces fit together.

jasonpang commented on extensions/amp-web-push/0.1/helper-frame.js:
```diff
@@ -0,0 +1,272 @@
+   * @param {string} queryString
+   * @return {!Object<string>}
+   */
+  parseQueryString(queryString) {
```
Nice catch, thanks!

jasonpang commented on extensions/amp-web-push/0.1/iframehost.js:
```diff
@@ -0,0 +1,86 @@
+  /**
+   * Returns the IFrame DOM element.
+   *
+   * @return {(Element|null)}
```
Corrected.

jasonpang commented on extensions/amp-web-push/0.1/iframehost.js:
```diff
@@ -0,0 +1,86 @@
+    return this.domElement_;
+  }
+
+  whenReady() {
```
Added JSDoc everywhere.

jasonpang commented on extensions/amp-web-push/0.1/permission-dialog.js:
```diff
@@ -0,0 +1,121 @@
+    if (options && options.debug) {
+      // Debug enables verbose logging for this page and the window and worker
+      // messengers
+      this.debug = true;
```
Fixed.

Changed the `if { .. }` block setting the boolean flag to just `this.debug_ = options && options.debug;` with the flag being private.

jasonpang commented on extensions/amp-web-push/0.1/permission-dialog.js:
```diff
@@ -0,0 +1,121 @@
+    });
+  }
+
+  /** @private */
```
Enhanced method JSDoc.

jasonpang commented on extensions/amp-web-push/0.1/permission-dialog.js:
```diff
@@ -0,0 +1,121 @@
+      window.opener !== window;
+  }
+
+  /** @private */
```
Enhanced method JSDoc.

dvoytenko commented:
@jasonpang Cool. Looks good and very close. I just made two comments: one more attempt at name bikeshedding and the issue with querying global `window.document`.

Thanks!

dvoytenko commented on extensions/amp-web-push/0.1/helper-frame.js:
```diff
@@ -0,0 +1,347 @@
+}
+
+if (!getMode().test) {
+  window.controller = new AmpWebPushHelperFrame({
```
Just wanted to ask if there's any concern of name collision?

jasonpang commented:
@dvoytenko Thanks for all your in-depth review suggestions so far! I've made your suggested changes:

- `<amp-web-push>` is now the config element, `<amp-web-push-widget>` is now the widget element
- Use `getAmpDoc()` instead of `win.document.createElement()` when checking for duplicate `<amp-web-push>` config elements
- Renamed global variable `controller` in `helper-frame.js` and `permission-dialog.js` to `_ampWebPushHelperFrame` and `_ampWebPushPermissionDialog` to avoid any potential naming conflicts. While it shouldn't happen at all since these are lightweight pages with only the script running on it, a unique name was a better choice.
- Also fixed some test race conditions

The last change we'd like to make is to host `helper-frame.js` and `permission-dialog.js` on `cdn.ampproject.org` so that each publisher file (`helper-frame.html` and `permission-dialog.html`) would include a `<script src="https://cdn.ampproject.org/v0/corresponding-file.js">` instead of the direct minified source. This allows for updates in the future.

@cramforce Could you take a second look at the changes so far?

honeybadgerdontcare commented on extensions/amp-web-push/validator-amp-web-push.protoascii:
```diff
@@ -0,0 +1,71 @@
+  }
+  attrs: {
+    name: "helper-iframe-url"
+    value_regex: "https:\/\/.*"
```
Instead of `value_regex` use `value_url`.

```
    value_url: {
      allowed_protocol: "https"
      allow_relative: false
    }
```

honeybadgerdontcare commented on extensions/amp-web-push/validator-amp-web-push.protoascii:
```diff
@@ -0,0 +1,71 @@
+  }
+  attrs: {
+    name: "permission-dialog-url"
+    value_regex: "https:\/\/.*"
```
Instead of `value_regex` use `value_url`.

```
    value_url: {
      allowed_protocol: "https"
      allow_relative: false
    }
```

honeybadgerdontcare commented on extensions/amp-web-push/validator-amp-web-push.protoascii:
```diff
@@ -0,0 +1,71 @@
+  }
+  attrs: {
+    name: "service-worker-url"
+    value_regex: "https:\/\/.*"
```
Instead of `value_regex` use `value_url`.

```
    value_url: {
      allowed_protocol: "https"
      allow_relative: false
    }
```

jasonpang commented:
@honeybadgerdontcare Fixed!

cramforce commented:
I'll defer to @dvoytenko for final approval. This looks super nice! Thanks for all the changes.

dvoytenko commented:
@jasonpang @cramforce Cool. I'm ready to approve then, so let me know if you're ready on your side.

RE:

> The last change we'd like to make is to host helper-frame.js and permission-dialog.js on cdn.ampproject.org so that each publisher file (helper-frame.html and permission-dialog.html) would include a <script src="https://cdn.ampproject.org/v0/corresponding-file.js"> instead of the direct minified source.

I recommend filing a bug for this for improvement so that we can merge this change first. It's definitely possible to deploy these to CDN, but we'd need an advice from @erwinmombay .

jasonpang commented:
@dvoytenko Sounds good, we'll tackle that feature later. 

I'm ready to merge!

honeybadgerdontcare commented:
This is now in production.
